design/ceph/multus-network.md

```diff
@@ -0,0 +1,129 @@
+- static: Allocate a static IPv4/IPv6 addresses to container and it's useful in debugging purpose.
+This cannot at scale because this means we will have to allocate IPs for **all** the daemon so it's not scalable.
+
+You can find a more detailed analysis at the end of the document in the [rejected proposal section](#rejected-proposals).
+
+## Ceph daemons implementation challenges
+
+### Monitors and OSDs
+
+The Ceph monitors only need access to the public network.
+The OSDs needs access to both public and cluster networks.
+
+Monitors requirements so far are the following:
+
+- Predictable IP addresses
+- Keep the same IP address for the life time of the deployment (IP should survive a restart)
+
+### RGW implementation
+
+Only need access to the public network.
+They use service IP with a load-balancer so we need to be careful.
+
+### MDS/RBD-MIRROR/NFS implementation
+
+Only need access to the public network.
+Nothing to do in particular since they don't use any service IPs.
+
+### CSI pods
+
+We can add annotations to these pods and they can reach out the Ceph public network, then the driver will expose the block or the filesystem normally.
+
+## Accepted proposal
+
+So far, the team has decided to go with the [whereabouts](https://github.com/dougbtv/whereabouts) IPAM.
+It's an IP Address Management (IPAM) CNI plugin that assigns IP addresses cluster-wide.
+If you need a way to assign IP addresses dynamically across your cluster -- Whereabouts is the tool for you. If you've found that you like how the host-local CNI plugin works, but, you need something that works across all the nodes in your cluster (host-local only knows how to assign IPs to pods on the same node) -- Whereabouts is just what you're looking for.
+Whereabouts can be used for both IPv4 & IPv6 addressing.
+
+It is under active development and is not ready but ultimately will allow to:
+
+- Have static IP addresses distributed across the cluster
+- These IPs will survive any deployment restart
+- The allocation will be done by whereabouts and not by Rook
+
+## Rejected proposals
+
+The following proposal were rejected but we keep them here for traceability and knowledge.
+
+### IPAM type 'DHCP'
+
+In this scenario, a DHCP server will distribute an IP address to a pod using a given range.
+
+Pros:
+
+- Pods will get a dedicated IP on a physical network interface
+- No changes required in Ceph, Rook will detect the CIDR via the `NetworkAttachmentDefinition` then populate Ceph's flag `public_network` and `cluster_network`
+
+Cons:
+
+- IP allocation not predictable, we don't know it until the pod is up and running.
+So the detection must append inside the monitor container is running, similarly to what the OSD code does today.
+- Requires drastic changes in the monitor bootstrap code
+- This adds a DHCP daemon on every part of the cluster and this has proven to be troublesome
```

Comment:
Ensuring a constant ip address is still the key point to be clear on. If it's using DHCP how can we ensure a stable ip address? Set the expiration on the ip to never expire? Or how can we ensure the ip address doesn't change at some point in the future when the mon pod is restarted?

Reply:
Currently, the DHCP does not support permanent lease AFAIK.